control-plane/controllers/configentries/terminatinggateway_controller.go

```diff
@@ -41,6 +116,221 @@ func (r *TerminatingGatewayController) UpdateStatus(ctx context.Context, obj cli
 	return r.Status().Update(ctx, obj, opts...)
 }
 
-func (r *TerminatingGatewayController) SetupWithManager(mgr ctrl.Manager) error {
+func (r *TerminatingGatewayController) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	// setup the index to lookup registrations by service name
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &v1alpha1.TerminatingGateway{}, terminatingGatewayByLinkedServiceName, termGWLinkedServiceIndexer); err != nil {
+		return err
+	}
+
 	return setupWithManager(mgr, &consulv1alpha1.TerminatingGateway{}, r)
 }
+
+func termGWLinkedServiceIndexer(o client.Object) []string {
+	termGW := o.(*v1alpha1.TerminatingGateway)
+	names := make([]string, 0, len(termGW.Spec.Services))
+	for _, service := range termGW.Spec.Services {
+		names = append(names, service.Name)
+	}
+
+	return names
+}
+
+func (r *TerminatingGatewayController) UpdateStatusFailedToSetACLs(ctx context.Context, termGW *consulv1alpha1.TerminatingGateway, err error) {
+	termGW.SetSyncedCondition(corev1.ConditionFalse, consulv1alpha1.TerminatingGatewayFailedToSetACLs, err.Error())
+	termGW.SetACLStatusConditon(corev1.ConditionFalse, consulv1alpha1.TerminatingGatewayFailedToSetACLs, err.Error())
+	if err := r.UpdateStatus(ctx, termGW); err != nil {
+		r.Log.Error(err, "error updating status")
+	}
+}
+
+func (r *TerminatingGatewayController) aclsEnabled() (bool, error) {
+	state, err := r.ConfigEntryController.ConsulServerConnMgr.State()
+	if err != nil {
+		return false, err
+	}
+	return state.Token != "", nil
+}
+
+func (r *TerminatingGatewayController) updateACls(log logr.Logger, termGW *consulv1alpha1.TerminatingGateway) error {
+	client, err := consul.NewClientFromConnMgr(r.ConfigEntryController.ConsulClientConfig, r.ConfigEntryController.ConsulServerConnMgr)
+	if err != nil {
+		return err
+	}
+
+	roles, _, err := client.ACL().RoleList(nil)
+	if err != nil {
+		return err
+	}
+
+	terminatingGatewayRoleID := ""
+	for _, role := range roles {
+		if strings.Contains(role.Name, termGW.Name) {
+			terminatingGatewayRoleID = role.ID
+			break
+		}
+	}
+
+	if terminatingGatewayRoleID == "" {
+		return errors.New("terminating gateway role not found")
+	}
+
+	termGwRole, _, err := client.ACL().RoleRead(terminatingGatewayRoleID, nil)
```

Comment:
#nit feels like we might as well go with `terminatingGatewayRole` based on the variable names above